### ISSUE TITLE
[BUG] [Formatter] H017 false positive

### ISSUE THREAD

**Issue by danielniccoli** [Riverside-Healthcare/djLint#372]
- [x ] I'm on the [latest version](https://pypi.org/project/djlint/) of djLint
- [x ] I've searched the [issues](https://github.com/Riverside-Healthcare/djLint/issues)
- [x ] I've read the [docs](https://djlint.com)

## System Info
 - OS: Windows 11
 - Python 3.10.6
 - djLint 1.12.3
 - template language: e.g. django

## Issue
djlint considers `<meta charset="utf-8">` a problem, although the syntax is correct. I'm going to quote the [MDN](https://developer.mozilla.org/en-US/docs/Glossary/Empty_element) on this one:

> An empty element is an [element](https://developer.mozilla.org/en-US/docs/Glossary/Element) from HTML, SVG, or MathML that **cannot have any child nodes** (i.e., nested elements or text nodes).
> 
> The [HTML](https://html.spec.whatwg.org/multipage/), [SVG](https://www.w3.org/TR/SVG2/), and [MathML](https://www.w3.org/TR/MathML3/) **specifications define very precisely what each element can contain**. Many combinations have no semantic meaning, for example an [\<audio>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/audio) element nested inside an [\<hr>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/hr) element.
> 
> In HTML, **using a closing tag on an empty element is usually invalid**. For example, `<input type="text"></input>` is invalid HTML

And for convencience, here's a link to the HTML specs that also shows meta being an empty element with some examples: https://html.spec.whatwg.org/multipage/semantics.html#the-head-element

## How To Reproduce
```html
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
```

**Comment by christopherpickering** [comment]
@danielniccoli again, this is a preference. You cited a reference saying they are not nessicary. There will most likely turn up someone who will disagree with you... which is why djlint will let you [disable](https://djlint.com/docs/configuration/) codes you don't agree with... like other linters also do 😉

**Comment by danielniccoli** [comment]
@christopherpickering You can disagree with an opinion, but if you disagree with the specification (a.k.a the facts), then you're being esoteric.

> Void elements only have a start tag; **end tags must not be specified for void elements**.

Source: [The HTML Specification](https://html.spec.whatwg.org/#void-elements)

**Comment by christopherpickering** [comment]
![](https://i.giphy.com/media/69mUSKBujnpgmxcqlg/giphy.webp)

**Comment by christopherpickering** [comment]
Just to be clear, rule **H017** is for self closing tags, suggesting to use `<meta />` in place of `<meta >` or `<img />` in place of `<img />`. I think the text you quoted is saying to **not** have a separate closing tag like `</meta>` or `</img>`, which I'm in agreement with.

**Comment by khink** [comment]
@christopherpickering How is this a preference? The specs that @danielniccoli provided are quite clear: `<meta ...>` is the way it should be done. It makes no sense for djlint to complain about that. 

I'm afraid that if i add an exception for H017, it will also not complain about tags that *should* be closed.

**Comment by cas--** [comment]
I have noticed this issue for a while and definitely feel it should be resolved. However it's not just meta it's all the void elements that should be a separate rule or just not flagged since there is a note of caution about marking a void element as self-closing:

From: https://html.spec.whatwg.org/#start-tags
> Then, if the element is one of the [void elements](https://html.spec.whatwg.org/#void-elements), or if the element is a [foreign element](https://html.spec.whatwg.org/#foreign-elements), then there may be a single U+002F SOLIDUS character (/), which on [foreign elements](https://html.spec.whatwg.org/#foreign-elements) marks the start tag as self-closing. On void elements it does not mark the start tag as self-closing but instead is unnecessary and has no effect of any kind. For such void elements, it should be used only with caution — especially since, if directly preceded by an [unquoted attribute value](https://html.spec.whatwg.org/#unquoted), it becomes part of the attribute value rather than being discarded by the parser.

**Comment by christopherpickering** [comment]
@cas-- I don't fully understand what you are recommending, h017 is only for void (self-closing) tags. @khink is suggesting that meta tags be split to their own rule.

What are you proposing as a solution?

**Comment by christopherpickering** [comment]
Probably the thing to do is make rules opt-in, such as #571 recommendation. I'm sure if that no-br rule showed up as a default, there would be a lot of hate mail 🤣

**Comment by christopherpickering** [comment]
I will let out a release with a few updates that will hopefully resolve this-

1. Rule H017 has been divided into a rule H017 and H035. H035 is exclusively for meta tags, and H017 is any other void tag.
2. Rule H017 name has been updated to say "void"
3. Rule H017 and H035 are disabled by default. They can be enabled with a new `--include` flag, or through the config file.

**Comment by khink** [comment]
Thanks for the fix and release! :pray:

**Comment by christopherpickering** [comment]
Hey @khink, @cas--, @danielniccoli Have any of you formatted w/ prettier before? How do you handle their conversion of `<img>` to `<img />`? I don't see a config option to disable it.

![image](https://user-images.githubusercontent.com/17788706/231825202-49d05ab4-4047-4b95-8dbe-66fa3725acca.png)

I'm planning to update the formatter to auto close these tags and wonder how you currently handle this. It's required for react html as well.

**Comment by christopherpickering** [comment]
https://github.com/prettier/prettier/issues/5246
I guess they don't have a flag :) I will plan to add one if I change this for the formatter.

**Comment by khink** [comment]
I haven't seen this. In the project i'm working on now, we only run prettier on JS/CSS. Possibly in other projects we'd run it on the React HTML, but not the Django templates.
